File: InheritanceTest/results/InheritanceTest.pl
use strict;
use warnings;

package BasePrinter;

sub new
{
    my $class = shift;
    my $self = {};
    bless $self, $class;
    return $self;
}

sub get_value {
    my ( $self ) = @_;
    return "Base";
}

sub print_it {
    my ( $self ) = @_;
    print(("BasePrinter: @{[$self->get_value()]}") . "\n");
}

sub some_base_func {
    my ( $self ) = @_;
    return 42;
}

package ChildPrinter;

use parent -norequire, 'BasePrinter';

sub new
{
    my $class = shift;
    my $self = {};
    bless $self, $class;
    return $self;
}

sub get_value {
    my ( $self ) = @_;
    return "Child";
}

package TestClass;

sub new
{
    my $class = shift;
    my $self = {};
    bless $self, $class;
    return $self;
}

sub get_printer {
    my ( $self, $name ) = @_;
    my $result = $name eq "child" ? new ChildPrinter() : new BasePrinter();
    return $result;
}

sub test_method {
    my ( $self ) = @_;
    my $base_printer = $self->get_printer("base");
    my $child_printer = $self->get_printer("child");
    $base_printer->print_it();
    $child_printer->print_it();
    print(("@{[$base_printer->some_base_func()]} == @{[$child_printer->some_base_func()]}") . "\n");
}

package Program;

eval {
    my $c = new TestClass();
    $c->test_method();
};
if ($@) {
    print "Exception: " . $@
}

File: InheritanceTest/results/InheritanceTest.rb
class BasePrinter
  attr_accessor(:num_value)

  def initialize()
      @num_value = 42
  end

  def get_value()
      return "Base"
  end

  def print_it()
      puts "BasePrinter: #{self.get_value()}"
  end

  def some_base_func()
      return self.num_value
  end
end

class ChildPrinter < BasePrinter
  def get_value()
      return "Child"
  end
end

class TestClass
  def get_printer(name)
      result = name == "child" ? ChildPrinter.new() : BasePrinter.new()
      return result
  end

  def test_method()
      base_printer = self.get_printer("base")
      child_printer = self.get_printer("child")
      base_printer.print_it()
      child_printer.print_it()
      puts "#{base_printer.some_base_func()} == #{child_printer.some_base_func()}"
      
      base_p2 = BasePrinter.new()
      child_p2 = ChildPrinter.new()
      puts "#{base_p2.num_value} == #{child_p2.num_value}"
  end
end

begin
    TestClass.new().test_method()
rescue Exception => err
    print "Exception: #{err}"
end